File: src/lib/radio/protocol.ts
export const FRAME_TYPE_FFT = 0x01;
export const FRAME_TYPE_AUDIO = 0x02;

export type RadioType = "sdr" | "transceiver";

export interface GainStage {
  name: string;
  label?: string;
  min: number;
  max: number;
  step: number;
}

export type RadioCommandCapability =
  | "tune"
  | "mode"
  | "gain"
  | "squelch"
  | "agc"
  | "antenna"
  | "filter"
  | "nr"
  | "nb"
  | "ptt"
  | "vfo"
  | "rit"
  | "xit"
  | "split"
  | "anf"
  | "qsk"
  | "vox"
  | "if_shift"
  | "cw_speed";

export interface RadioCapabilities {
  can_transmit: boolean;
  can_stream_iq: boolean;
  can_stream_fft: boolean;
  can_stream_audio: boolean;
  antennas: string[];
  modes: string[];
  frequency_range: [number, number];
  sample_rates: number[];
  gain_stages: GainStage[];
  /** Optional command-level feature negotiation. Missing advanced flags are unsupported. */
  commands?: Partial<Record<RadioCommandCapability, boolean>>;
}

export interface DeviceInfo {
  device_id: string;
  name: string;
  driver: string;
  type: RadioType;
  serial?: string;
  port?: string;
  available: boolean;
  capabilities: RadioCapabilities;
}

export interface RadioFilter {
  low: number;
  high: number;
}

export interface RadioNr {
  enabled: boolean;
  level: number;
}

export interface RadioNb {
  enabled: boolean;
  threshold?: number;
}

export interface RadioState {
  connected: boolean;
  freq: number;
  mode: string;
  antenna?: string;
  vfo?: "A" | "B";
  gains: Record<string, number>;
  agc: boolean;
  agcMode?: number;
  ptt?: boolean;
  filter?: RadioFilter;
  nr?: RadioNr;
  nb?: RadioNb;
  squelch?: number;
  signal_dbm?: number;
  rit?: { enabled: boolean; offsetHz: number };
  xit?: { enabled: boolean; offsetHz: number };
  split?: boolean;
  anf?: boolean;
  qsk?: boolean;
  vox?: boolean;
  lock?: boolean;
  txAntenna?: string;
  txMeter?: { powerW?: number; swr?: number; alc?: number };
  cwSpeed?: number;
  ifShift?: number;
}

export interface DaemonHelloMessage {
  type: "hello";
  version: string;
  daemon_id: string;
  features?: string[];
}

export interface DaemonResponseMessage {
  type: "response";
  id: string;
  success: boolean;
  error?: string;
}

export interface DevicesListMessage {
  type: "devices:list";
  devices: DeviceInfo[];
}

export interface DevicesAddedMessage {
  type: "devices:added";
  device_id: string;
  name: string;
}

export interface DevicesRemovedMessage {
  type: "devices:removed";
  device_id: string;
}

export interface RadioStateMessage {
  type: "radio:state";
  device_id: string;
  state: RadioState;
}

export interface RadioSmeterMessage {
  type: "radio:smeter";
  device_id: string;
  dbm: number;
}

export interface DaemonStatusMessage {
  type: "daemon:status";
  version: string;
  uptime_secs: number;
  platform: string;
  connected_radios: number;
  active_streams: number;
  cpu_percent: number;
  memory_mb: number;
}

export interface DaemonDiscoveryDaemonsMessage {
  type: "discovery:daemons";
  daemons: Array<{
    fullname: string;
    hostname: string;
    port: number;
    addresses: string[];
    txt: Record<string, string>;
  }>;
}

export interface WsjtxStatus {
  frequency: number;
  mode: string;
  dxCall?: string;
  dxGrid?: string;
  txEnabled: boolean;
  decoding: boolean;
  rxDF: number;
  txDF: number;
  instanceId?: string;
}

export interface WsjtxDecode {
  isNew: boolean;
  time: number;
  snr: number;
  deltaTime: number;
  deltaFrequency: number;
  mode: string;
  message: string;
  lowConfidence: boolean;
  callsign?: string;
  grid?: string;
  instanceId?: string;
}

export interface WsjtxQsoLogged {
  callsign: string;
  grid?: string;
  frequency: number;
  mode: string;
  reportSent: string;
  reportReceived: string;
  txPower: string;
  comments: string;
  timestamp: string;
  adifRecord?: string;
  instanceId?: string;
}

export interface WsjtxStatusMessage {
  type: "wsjtx:status";
  status: WsjtxStatus;
}

export interface WsjtxDecodeMessage {
  type: "wsjtx:decode";
  decode: WsjtxDecode;
}

export interface WsjtxQsoLoggedMessage {
  type: "wsjtx:qso_logged";
  qso: WsjtxQsoLogged;
}

export interface WsjtxClearMessage {
  type: "wsjtx:clear";
  window?: string;
  instanceId?: string;
}

export interface ClusterSpot {
  id?: string;
  key?: string;
  spotter: string;
  spotterGrid?: string;
  dx: string;
  dxGrid?: string;
  /** Frequency in kHz */
  freq: number;
  comment: string;
  time: string;
  mode?: string;
  band?: string;
}

export interface ClusterSpotMessage {
  type: "cluster:spot";
  spotter: string;
  spotterGrid?: string;
  dx: string;
  dxGrid?: string;
  /** Frequency in kHz */
  freq: number;
  comment: string;
  time: string;
  mode?: string;
  band?: string;
  id?: string;
  key?: string;
}

export interface ClusterStatusMessage {
  type: "cluster:status";
  node?: string;
  connected: boolean;
  spotsReceived: number;
  lastSpotTime?: string;
  key?: string;
}

export type DaemonIncomingMessage =
  | DaemonHelloMessage
  | DaemonResponseMessage
  | DevicesListMessage
  | DevicesAddedMessage
  | DevicesRemovedMessage
  | RadioStateMessage
  | RadioSmeterMessage
  | DaemonStatusMessage
  | DaemonDiscoveryDaemonsMessage
  | WsjtxStatusMessage
  | WsjtxDecodeMessage
  | WsjtxQsoLoggedMessage
  | WsjtxClearMessage
  | ClusterSpotMessage
  | ClusterStatusMessage
  | { type: string; [k: string]: unknown };

export function isDevicesListMessage(
  msg: DaemonIncomingMessage,
): msg is DevicesListMessage {
  return (
    msg.type === "devices:list" &&
    Array.isArray((msg as DevicesListMessage).devices)
  );
}

export function isDevicesAddedMessage(
  msg: DaemonIncomingMessage,
): msg is DevicesAddedMessage {
  return (
    msg.type === "devices:added" &&
    typeof (msg as DevicesAddedMessage).device_id === "string"
  );
}

export function isDevicesRemovedMessage(
  msg: DaemonIncomingMessage,
): msg is DevicesRemovedMessage {
  return (
    msg.type === "devices:removed" &&
    typeof (msg as DevicesRemovedMessage).device_id === "string"
  );
}

export function isRadioStateMessage(
  msg: DaemonIncomingMessage,
): msg is RadioStateMessage {
  return (
    msg.type === "radio:state" &&
    typeof (msg as RadioStateMessage).device_id === "string" &&
    typeof (msg as RadioStateMessage).state === "object"
  );
}

export function isRadioSmeterMessage(
  msg: DaemonIncomingMessage,
): msg is RadioSmeterMessage {
  return (
    msg.type === "radio:smeter" &&
    typeof (msg as RadioSmeterMessage).device_id === "string" &&
    typeof (msg as RadioSmeterMessage).dbm === "number"
  );
}

export function isDaemonStatusMessage(
  msg: DaemonIncomingMessage,
): msg is DaemonStatusMessage {
  return (
    msg.type === "daemon:status" &&
    typeof (msg as DaemonStatusMessage).version === "string" &&
    typeof (msg as DaemonStatusMessage).uptime_secs === "number"
  );
}

export function isDaemonResponseMessage(
  msg: DaemonIncomingMessage,
): msg is DaemonResponseMessage {
  return (
    msg.type === "response" &&
    typeof (msg as DaemonResponseMessage).id === "string" &&
    typeof (msg as DaemonResponseMessage).success === "boolean"
  );
}

export function isDaemonDiscoveryDaemonsMessage(
  msg: DaemonIncomingMessage,
): msg is DaemonDiscoveryDaemonsMessage {
  return (
    msg.type === "discovery:daemons" &&
    Array.isArray((msg as DaemonDiscoveryDaemonsMessage).daemons)
  );
}

export function isWsjtxStatusMessage(
  msg: DaemonIncomingMessage,
): msg is WsjtxStatusMessage {
  return (
    msg.type === "wsjtx:status" &&
    typeof (msg as WsjtxStatusMessage).status === "object"
  );
}

export function isWsjtxDecodeMessage(
  msg: DaemonIncomingMessage,
): msg is WsjtxDecodeMessage {
  return (
    msg.type === "wsjtx:decode" &&
    typeof (msg as WsjtxDecodeMessage).decode === "object"
  );
}

export function isClusterSpotMessage(
  msg: DaemonIncomingMessage,
): msg is ClusterSpotMessage {
  return (
    msg.type === "cluster:spot" &&
    typeof (msg as ClusterSpotMessage).freq === "number"
  );
}

export type RadioBinaryFrame =
  | {
      kind: "fft";
      devIdx: number;
      centerHz: number;
      spanHz: number;
      bins: Float32Array;
    }
  | { kind: "audio"; devIdx: number; sampleRate: number; samples: Int16Array };

export function generateCommandId(): string {
  return `${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
}

export function parseBinaryFrame(data: ArrayBuffer): RadioBinaryFrame | null {
  if (data.byteLength < 2) return null;
  const dv = new DataView(data);
  const frameType = dv.getUint8(0);
  const devIdx = dv.getUint8(1);

  if (frameType === FRAME_TYPE_FFT) {
    if (data.byteLength < 18) return null;
    if ((data.byteLength - 18) % 4 !== 0) return null;
    const centerHz = dv.getFloat64(2, true);
    const spanHz = dv.getFloat64(10, true);
    const bins = new Float32Array(data.slice(18));
    return { kind: "fft", devIdx, centerHz, spanHz, bins };
  }

  if (frameType === FRAME_TYPE_AUDIO) {
    if (data.byteLength < 6) return null;
    const sampleRate = dv.getUint32(2, true);
    const samples = new Int16Array(data, 6);
    return { kind: "audio", devIdx, sampleRate, samples };
  }

  return null;
}

File: daemon/propulse-daemon/src/protocol.rs
use serde::{Deserialize, Serialize};

use propulse_radio::types::{DeviceInfo, RadioState};

pub const PROTOCOL_VERSION: &str = "1.1.0";
pub const FRAME_TYPE_FFT: u8 = 0x01;
pub const FRAME_TYPE_AUDIO: u8 = 0x02;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Hello {
  #[serde(rename = "type")]
  pub kind: String,
  pub version: String,
  pub daemon_id: String,
  pub features: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DevicesList {
  #[serde(rename = "type")]
  pub kind: String,
  pub devices: Vec<DeviceInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RadioStateEvent {
  #[serde(rename = "type")]
  pub kind: String,
  pub device_id: String,
  pub state: RadioState,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RadioSmeterEvent {
  #[serde(rename = "type")]
  pub kind: String,
  pub device_id: String,
  pub dbm: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DaemonStatusEvent {
  #[serde(rename = "type")]
  pub kind: String,
  pub version: String,
  pub uptime_secs: u64,
  pub platform: String,
  pub connected_radios: usize,
  pub active_streams: usize,
  pub cpu_percent: f32,
  pub memory_mb: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Response {
  #[serde(rename = "type")]
  pub kind: String,
  pub id: String,
  pub success: bool,
  #[serde(skip_serializing_if = "Option::is_none")]
  pub error: Option<String>,
}

impl Response {
  pub fn ok(id: impl Into<String>) -> Self {
    Self {
      kind: "response".to_string(),
      id: id.into(),
      success: true,
      error: None,
    }
  }

  pub fn err(id: impl Into<String>, message: impl Into<String>) -> Self {
    Self {
      kind: "response".to_string(),
      id: id.into(),
      success: false,
      error: Some(message.into()),
    }
  }
}

pub fn build_fft_frame(dev_idx: u8, center_hz: f64, span_hz: f64, bins: &[f32]) -> Vec<u8> {
  let mut out = Vec::with_capacity(1 + 1 + 8 + 8 + (bins.len() * 4));
  out.push(FRAME_TYPE_FFT);
  out.push(dev_idx);
  out.extend_from_slice(&center_hz.to_le_bytes());
  out.extend_from_slice(&span_hz.to_le_bytes());
  for v in bins {
    out.extend_from_slice(&v.to_le_bytes());
  }
  out
}

pub fn build_audio_frame(dev_idx: u8, sample_rate: u32, samples: &[i16]) -> Vec<u8> {
  let mut out = Vec::with_capacity(1 + 1 + 4 + (samples.len() * 2));
  out.push(FRAME_TYPE_AUDIO);
  out.push(dev_idx);
  out.extend_from_slice(&sample_rate.to_le_bytes());
  for s in samples {
    out.extend_from_slice(&s.to_le_bytes());
  }
  out
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn fft_frame_roundtrip() {
    let dev_idx = 7;
    let center = 14_074_000.0;
    let span = 2_048_000.0;
    let bins = [1.0f32, -5.5f32, 3.25f32];

    let bytes = build_fft_frame(dev_idx, center, span, &bins);
    assert_eq!(bytes[0], FRAME_TYPE_FFT);
    assert_eq!(bytes[1], dev_idx);

    let center_bytes: [u8; 8] = bytes[2..10].try_into().unwrap();
    let span_bytes: [u8; 8] = bytes[10..18].try_into().unwrap();
    assert_eq!(f64::from_le_bytes(center_bytes), center);
    assert_eq!(f64::from_le_bytes(span_bytes), span);

    let b0: [u8; 4] = bytes[18..22].try_into().unwrap();
    let b1: [u8; 4] = bytes[22..26].try_into().unwrap();
    let b2: [u8; 4] = bytes[26..30].try_into().unwrap();
    assert_eq!(f32::from_le_bytes(b0), bins[0]);
    assert_eq!(f32::from_le_bytes(b1), bins[1]);
    assert_eq!(f32::from_le_bytes(b2), bins[2]);
  }

  #[test]
  fn audio_frame_roundtrip() {
    let dev_idx = 1;
    let sample_rate = 48_000u32;
    let samples = [0i16, 1234i16, -1234i16, i16::MAX];

    let bytes = build_audio_frame(dev_idx, sample_rate, &samples);
    assert_eq!(bytes[0], FRAME_TYPE_AUDIO);
    assert_eq!(bytes[1], dev_idx);

    let sr: [u8; 4] = bytes[2..6].try_into().unwrap();
    assert_eq!(u32::from_le_bytes(sr), sample_rate);

    let s0: [u8; 2] = bytes[6..8].try_into().unwrap();
    let s1: [u8; 2] = bytes[8..10].try_into().unwrap();
    let s2: [u8; 2] = bytes[10..12].try_into().unwrap();
    let s3: [u8; 2] = bytes[12..14].try_into().unwrap();
    assert_eq!(i16::from_le_bytes(s0), samples[0]);
    assert_eq!(i16::from_le_bytes(s1), samples[1]);
    assert_eq!(i16::from_le_bytes(s2), samples[2]);
    assert_eq!(i16::from_le_bytes(s3), samples[3]);
  }
}
